### ISSUE TITLE
Dockerd slows down and finally hangs

### ISSUE THREAD

**Issue by aartur** [moby/moby#43870]
### Description

`dockerd` is running on EC2 `m5n.4xlarge` 64GB/16VCPU instance. Containers are launched by Nomad 1,3,1 at the rate of about 2-3 per second, keeping about 100 containers running at the given time.

The problem is that it takes more and more time for `dockerd` to launch a container. After 12-36 hours of `dockerd` uptime most of Nomad allocations are in the `STARTING` state for a few minutes (and after restarting `dockerd`, they stay only for 0-2 seconds in this state). And usually after a few more hours `dockerd` hangs completely - no new containers are started, `docker info` command hangs.

What I tried (without any positive effect):
- tuning `sysctl` settings (enlarging Unix socket queues, enlarging socket/fs/watches/pipes limits)
- downgrading `containerd` to 1.5
- disabling log collection from `dockerd` by Nomad
- changing `dockerd` log driver to `journald`

### Reproduce

1. Run Nomad client node on `m5n.4xlarge`
2. Submit Nomad allocations (that use the default `docker` driver) to the client node at the rate of about 2-3 per second, each allocation consuming about 300MB of RAM and running for 30-300 seconds.
3. Let the node run for 12-48 hours
4. Check that `dockerd` is hanging, e.g. `docker info` command will hang, and logs indicate no new containers are being started.

### Expected behavior

`dockerd` shouldn't ever hang.

### docker version

```bash
Client: Docker Engine - Community
 Version:           20.10.17
 API version:       1.41
 Go version:        go1.17.11
 Git commit:        100c701
 Built:             Mon Jun  6 23:02:46 2022
 OS/Arch:           linux/amd64
 Context:           default
 Experimental:      true

Server: Docker Engine - Community
 Engine:
  Version:          20.10.17
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.17.11
  Git commit:       a89b842
  Built:            Mon Jun  6 23:00:51 2022
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.6.6
  GitCommit:        10c12954828e7c7c9b6e0ea9b0c02b01407d3ae1
 runc:
  Version:          1.1.2
  GitCommit:        v1.1.2-0-ga916309
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

### docker info

```bash
This command hangs, based on `strace` it waits on "GET /v1.41/info" result.
```

### Additional Info

The EC2 instance runs Ubuntu 22.04 with the latest `linux-aws` kernel `5.15.0-1015-aws`, Nomad 1.3.1, and the latest Docker+Containerd installed from APT repository at https://download.docker.com/linux/ubuntu.

Goroutine stacks:
[goroutine-stacks-2022-07-26T182207Z.log.gz](https://github.com/moby/moby/files/9192792/goroutine-stacks-2022-07-26T182207Z.log.gz)

While the `dockerd` is hanging, CPU and disk usage is close to zero. Latest `docker` debug logs contain mostly `Calling POST /containers/2edbe69.../start` lines.

`dmesg` contains a few warnings:
```
[24959.511473] overlayfs: lowerdir is in-use as upperdir/workdir of another mount, accessing files from both mounts will result in undefined behavior.
```
(but the rate of these messages is low, maybe one per 2000 of launched containers).

`docker ps` and `docker image ls` commands are not hanging, they produce output.

The other most similar issue I found is https://github.com/moby/moby/issues/42682#issuecomment-1066198793.

**Comment by olljanat** [comment]
Just guessing that this also might be related to kernel version like #43755

Is there AWS instances available with older kernel version? If so it would be interesting to known if behavior is same with those?

**Comment by aartur** [comment]
@olljanat I checked your suggestion and it looks like it fixed the problems, thank you.

I downgraded the instance from Ubuntu 22.04 to Ubuntu 20.04 to use an older kernel (I didn't find a way to install an older AWS kernel on 22.04) and `dockerd` works well for 48+ hours now, without any slow-downs or hangs. On 22.04, it was consistently very slow or hanging after 48 hours of `dockerd` uptime.

I downgraded the kernel on 20.04 by installing `linux-aws-lts-20.04` which currently provides `5.4.0-1081-aws`. From what I see it also downgraded `cgroups` from `v2` to `v1`.

This is `docker info` output from an instance that uses the problematic Ubuntu 22.04 config:
```
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  app: Docker App (Docker Inc., v0.9.1-beta3)
  buildx: Docker Buildx (Docker Inc., v0.8.2-docker)
  scan: Docker Scan (Docker Inc., v0.17.0)

Server:
 Containers: 9
  Running: 9
  Paused: 0
  Stopped: 0
 Images: 10
 Server Version: 20.10.17
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: systemd
 Cgroup Version: 2
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 io.containerd.runtime.v1.linux runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 10c12954828e7c7c9b6e0ea9b0c02b01407d3ae1
 runc version: v1.1.2-0-ga916309
 init version: de40ad0
 Security Options:
  apparmor
  seccomp
   Profile: default
  cgroupns
 Kernel Version: 5.15.0-1014-aws
 Operating System: Ubuntu 22.04 LTS
 OSType: linux
 Architecture: x86_64
 CPUs: 16
 Total Memory: 30.57GiB
 Name: ip-10-2-250-164
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
```

And this is `docker info` output for the Ubuntu 20.04 instance:
```
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  app: Docker App (Docker Inc., v0.9.1-beta3)
  buildx: Docker Buildx (Docker Inc., v0.8.2-docker)
  scan: Docker Scan (Docker Inc., v0.17.0)

Server:
 Containers: 198
  Running: 197
  Paused: 0
  Stopped: 1
 Images: 1
 Server Version: 20.10.17
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Cgroup Version: 1
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runtime.v1.linux runc io.containerd.runc.v2
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 10c12954828e7c7c9b6e0ea9b0c02b01407d3ae1
 runc version: v1.1.2-0-ga916309
 init version: de40ad0
 Security Options:
  apparmor
  seccomp
   Profile: default
 Kernel Version: 5.4.0-1081-aws
 Operating System: Ubuntu 20.04.4 LTS
 OSType: linux
 Architecture: x86_64
 CPUs: 16
 Total Memory: 62.03GiB
 Name: ip-10-2-243-13
 Docker Root Dir: /var/lib/docker
 Debug Mode: true
  File Descriptors: 2590
  Goroutines: 2802
  System Time: 2022-07-29T07:33:03.867098905Z
  EventsListeners: 0
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
```

**Comment by thaJeztah** [comment]
Perhaps someone is able to try to dig into the issue to see if we can narrow down what's causing it

@aartur do you have minimal reproduction steps to reproduce the issue without Nomad?

**Comment by aartur** [comment]
I was able to reproduce the problem (Docker daemon hanging) by just keeping a pool of 200 containers running `sleep`:

```
$ seq 100000000 | parallel -j 200 docker run --network host archlinux '/bin/bash -c "sleep 3 && echo Job {} finished"'
```

The final lines before the hang, it took about 3 hours for this to happen:
```
Job 83412 finished
Job 83415 finished
Job 83416 finished
Job 83417 finished
Job 83422 finished
Job 83418 finished
```

I used `--network host` to speed up initialization and it's also what our Nomad jobs are mostly using.

The command was run on Ubuntu 22.04 with `docker-ce==5:20.10.21~3-0~ubuntu-jammy` and kernel `5.15.0-52-generic`.

And my previous comment mentioning that downgrading to Ubuntu 20.04+cgroups v1 fixed the problems was actually not valid, we see Docker daemon hanging also using this setup, but the occurrence rate seems much lower. We also saw the hangs using Centos Stream 9 + cgroups v1, but with a occurrence rate similar to Ubuntu 20.04.
